src/Domain/DonationRepository.php

```diff
@@ -455,6 +455,78 @@ public function findWithTransferIdInArray(array $transferIds): array
         return $donations;
     }
 
+    /**
+     * Taking the floor of the current minute as N and looking between N-16 minutes and
+     * N-1 minutes for donations with >£0 matching assigned, returns:
+     * * if there are less than 20 such donations, null; or
+     * * if there are 20+ such donations, the ratio of those which are complete.
+     */
+    public function getRecentHighVolumeCompletionRatio(\DateTimeImmutable $startOfThisMinute): ?float
+    {
+        $oneMinuteAgo = $startOfThisMinute->sub(new \DateInterval('PT1M'));
+        $sixteenMinutesAgo = $startOfThisMinute->sub(new \DateInterval('PT16M'));
+
+        $qb = $this->getEntityManager()->createQueryBuilder()
+            ->select(
+                'COUNT(d.id) as donationCount, SUM(CASE WHEN d.donationStatus IN ' .
+                '(:completeStatuses) THEN 1 ELSE 0 END) as completeCount'
+            )
+            ->from(Donation::class, 'd')
+            ->leftJoin('d.fundingWithdrawals', 'fw')
+            ->where('d.createdAt >= :start')
+            ->andWhere('d.createdAt < :end')
+            ->having('SUM(fw.amount) > 0')
+            ->setParameter('start', $sixteenMinutesAgo)
+            ->setParameter('end', $oneMinuteAgo)
+            ->setParameter(
+                'completeStatuses',
+                array_map(static fn(DonationStatus $s) => $s->value, DonationStatus::SUCCESS_STATUSES),
+            );
+
+        /**
+         * @var array{donationCount: int, completeCount: int}|null $result
+         */
+        $result = $qb->getQuery()->getOneOrNullResult(Query::HYDRATE_ARRAY);
+
+        if ($result === null || $result['donationCount'] < 20) {
+            return null;
+        }
+
+        return (float) $result['completeCount'] / $result['donationCount'];
+    }
+
+    public function getDonationsJustCreated(\DateTimeImmutable $startOfThisMinute): int
```

Comment:
might be slightly clearer named without assuming the given minute is now, and also count instead of get? e.g. 
```suggestion
    public function countDonationsCreatedInMinuteTo(\DateTimeImmutable $minuteEnd): int
```

The given param is the end of the period of interest, not the beginning. (true its also going to be the start of the current minute at runtime, but repo code doesn't need to know that)

Same applies to the following function.

Reply:
Updated some var names in 4bda1919d6407cc86ef89efec3cc6f1f1da62d67